lua/pl/utils.lua

```diff
@@ -618,36 +618,42 @@ end
 -- An application can override this function to support proper output of
 -- deprecation warnings. The warnings can be generated from libraries or
 -- functions by calling `utils.raise_deprecation`. By default this function
--- doesn't do anything.
+-- will write to the 'warn' system (introduced in Lua 5.4, or the compatibility
+-- function from the `compat` module for earlier versions).
 --
 -- Note: only applications should override this function, libraries should not.
 -- @string msg the message to display/log
 -- @string trace the traceback from where the deprecated element was invoked
 -- @usage
+-- -- write to the Nginx logs with OpenResty
 -- function utils.deprecation_warning(msg, trace)
---   io.stderr:write(msg .. "\n" .. trace .."\n")
+--   ngx.log(ngx.WARN, msg, " ", trace)
+-- end
+--
+-- -- disable deprecation warnings
+-- function utils.deprecation_warning(msg, trace)
+--   -- do nothing
 -- end
 function utils.deprecation_warning(msg, trace)
-  -- this does nothing by default
+  warn(msg, "\n", trace)  -- luacheck: ignore
```

Comment:
@alerque need some input here; since this PR adds a compatibility function `warn`, do we then still need this overridable function `utils.deprecation_warning` or should we directly call `warn` from the `raise_deprecation` function below and let the user override the `warn` function?

Seems they serve the same purpose, so the extra level of indirection is not needed.

Reply:
The de serve roughly the same purpose, but I think the extra level of abstraction could prove useful. The `warn()` function being an integral part of the language means it may be used by other libraries and/or overridden for other reasons. It's okay that we pass most of our work onto it, but I think we should continue using `utils.deprecation_warning()` internally to allow for handling Penlight related messages differently than how `warn()` works. I don't want people to get into a situation where they want to discard Penlight warnings but keep warnings from their own app, but since they used `warn()` in their own app and we also used it directly in Penlight there is no longer any way for them to separate the handling. If we use our own function internally it gives people the possibility to handle them differently.

Reply:
Yes that was my doubt as well. But... if the application uses more than 1 library, it already has to deal with the same issue, an all-or-nothing choice. So either every library has to add an extra level of abstraction, or otherwise it's only a half baked solution... From that perspective I'm inclined to just drop it here as well...

Does that make sense?

Reply:
Given Penlight's role as a kind of standard library setting up base level functionality for the rest of what might be built on top, no I don't think it does make sense to just assume nobody will find this useful. It's also quite early in the game and we don't know what other libraries may or may not play along. Frankly I doubt adoption of `warn()` will even happen any time soon for most libraries. I'm not inclined to throw in the towel without even trying. It's a super low cost operation in a vary rare use circumstance that adds a significant level of flexibility. Seems worth it to me even if no other libraries are involved just to free up the user to reserve Lua's functionality for themselves without us cluttering it up if so desired.